src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2241,143 @@ int64_t BlueFS::_read_random(
   return ret;
 }
 
+void BlueFS::_wal_update_size(FileRef file) {
+  ceph_assert(file->is_wal_read_loaded);
+  uint64_t current_size = file->fnode.size;
+  uint64_t flush_offset = current_size;
+
+  FileReader *h = new FileReader(file, cct->_conf->bluefs_max_prefetch, false, true);
+
+  while(flush_offset < file->fnode.allocated) {
+    // read first part of wal flush
+    bufferlist bl;
+    _read(h, flush_offset, sizeof(File::WALFlush::WALLength), &bl, nullptr);
```

Comment:
This is bad.
FileReader will redirect reading to _read_wal, that is already trying to cut envelope out.

Reply:
Here I treat `_read` as `read raw data`, `fnode.size` includes all extra envelope data. `BlueFS::read` is the one that redirect to `_read` or `_read_wal`.